### PR TITLE
MM-16082, MM-16382, MM-17057 - Updating a11y navigation post create

### PR DESCRIPTION
**Pull request by asaadmahmood**
#### Summary
MM-16082 - Updating a11y navigation post create
MM-16382 - Updating a11y navigation comment create
MM-17057 - Fixes upload menu bug

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-16082
https://mattermost.atlassian.net/browse/MM-16382
https://mattermost.atlassian.net/browse/MM-17057

### CODE REVIEW

File: components/create_comment/create_comment.jsx [hunk 1]
```diff
@@ -266,6 +267,10 @@ export default class CreateComment extends React.PureComponent {
         }
     }
 
+    updateState = (newState) => {
```
Comment (crspeller) [inline components/create_comment/create_comment.jsx]: Probably clearer to rename this function to updatePreview

File: components/create_post/create_post.jsx [hunk 2]
```diff
@@ -324,6 +321,10 @@ export default class CreatePost extends React.Component {
         this.removeOrientationListeners();
     }
 
+    updateState = (newState) => {
```
Comment (crspeller) [inline components/create_post/create_post.jsx]: Same comment as above.

File: components/edit_post_modal/edit_post_modal.jsx [hunk 3]
```diff
@@ -59,6 +65,10 @@ export default class EditPostModal extends React.PureComponent {
         }
     }
 
+    updateState = (newState) => {
```
Comment (crspeller) [inline components/edit_post_modal/edit_post_modal.jsx]: Same as above.

File: components/file_upload/file_upload.jsx [hunk 4]
```diff
@@ -635,8 +650,13 @@ export default class FileUpload extends PureComponent {
                 );
             });
             bodyAction = (
-                <React.Fragment>
+                <button
```
Comment (crspeller) [inline components/file_upload/file_upload.jsx]: Maybe I just don't understand what is going on here. But why do we have a button wrapping a file input field? Why does this have a separate onClick handler that checks if files can be uploaded?

File: components/post_view/post_list/post_list_virtualized.jsx [hunk 5]
```diff
@@ -482,15 +482,14 @@ export default class PostList extends React.PureComponent {
...
-                            role='presentation'
                             aria-live='polite'
                             id='postListContent'
                             className='post-list__content'
                         >
                             <AutoSizer>
                                 {({height, width}) => (
                                     <DynamicSizeList
-                                        role='presentation'
+                                        role='”listbox'
```
Comment (crspeller) [inline components/post_view/post_list/post_list_virtualized.jsx]: Strange backwards quotes intentional?

File: components/quick_input.jsx [hunk 6]
```diff
@@ -9,6 +9,11 @@ import React from 'react';
 export default class QuickInput extends React.PureComponent {
     static propTypes = {
 
+        /**
+         * Aria label for the textarea
+         */
+        ariaLabel: PropTypes.string,
```
Comment (crspeller) [inline components/quick_input.jsx]: This doesn't look like it's used.

File: components/file_upload/file_upload.jsx (hunk 4 above)
Comment (asaadmahmood) [reply components/file_upload/file_upload.jsx]: The button is wrapping the `input file` button.
The `input file` button is hidden as we do want it to be seen, thus it is not visible for users visually, but is still present.

We can't show a visual navigation on the `input file` button as that would mean making it visible, which we do not want to do, thus we are wrapping

The button allows us to navigate on it, we can make it a div as well with a `role="button"` but I think this is better.

Then when a user presses `enter` on the `button` (via keyboard navigation), we check if a user can upload a file, and then proceed to uploading it.

Comment (asaadmahmood) [reply components/file_upload/file_upload.jsx]: However, I don't think we need to wrap the input in a button to do that, so I've separated the button out as a child, and also removed the extra function.

File: components/textbox/textbox_links.jsx [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+import * as Utils from 'utils/utils.jsx';
+
+const PreReleaseFeatures = Constants.PRE_RELEASE_FEATURES;
+
+export default class TextboxLinks extends React.Component {
+    static propTypes = {
+        preview: PropTypes.bool,
+        characterLimit: PropTypes.number.isRequired,
+        previewMessageLink: PropTypes.string,
+        updateState: PropTypes.func,
```
Comment (saturninoabril) [inline components/textbox/textbox_links.jsx]: Should this be `updatePreview`?

File: components/textbox/textbox_links.jsx [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+                    className='style--none textbox-preview-link color--link'
+                >
+                    {this.props.preview ? (
+                        editHeader
+                    ) : (
+                        <FormattedMessage
+                            id='textbox.preview'
+                            defaultMessage='Preview'
+                        />
+                    )}
```
Comment (saturninoabril) [inline components/textbox/textbox_links.jsx]: This conditional is confusing.  It's like: if preview, show edit header; if not preview, then show preview.

File: components/create_post/create_post.jsx [hunk 9]
```diff
@@ -50,11 +51,6 @@ function trimRight(str) {
 export default class CreatePost extends React.Component {
     static propTypes = {
 
-        /**
-         *  ref passed from channelView for EmojiPickerOverlay
-         */
-        getChannelView: PropTypes.func,
```
Comment (saturninoabril) [inline components/create_post/create_post.jsx]: Is this related to this PR? It's good to clean it up here but might be better to deal with it separately as there's other place that needs to remove as well.

File: components/textbox/textbox_links.jsx (hunk 8 above)
Comment (saturninoabril) [reply components/textbox/textbox_links.jsx]: nvm and thanks for checking.

File: components/create_post/create_post.jsx (hunk 9 above)
Comment (asaadmahmood) [reply components/create_post/create_post.jsx]: Done.

File: components/textbox/textbox_links.jsx (hunk 7 above)
Comment (asaadmahmood) [reply components/textbox/textbox_links.jsx]: Done.

**Comment by crspeller** [comment]
@asaadmahmood Looks good but worried about breaks given the stability of some of the other accessibility code. Adding a QA test requirement.

**Comment by lindy65** [comment]
Thanks @asaadmahmood 

1. For https://mattermost.atlassian.net/browse/MM-17057, tested file and image uploads in the center panel and RHS and did not find any issues. Also tested dragging and dropping images and files to center panel and RHS and did not find any issues.

2. Testing https://mattermost.atlassian.net/browse/MM-16082 
    - the read-out is not reading what is in 1) of the ticket description. It is saying "Message input *complimentary region*, Write to [channel name] edit type in text" not "Message input write to [Channel name] edit type in text".
    - For 4) of the ticket description the read-out is saying "Preview *button*" not "Preview *link*

3. Testing https://mattermost.atlassian.net/browse/MM-16382
    - For 4) in the ticket description the read-out is saying "Preview *button*" not "Preview link"

**Comment by asaadmahmood** [comment]
Thanks.
2. This only reproduces if you're coming off from a different section, if you just navigate in between that section, it should just say:
"Message input write to [Channel name] edit type in text"
So that's expected.

3. That's expected too, as the Preview link is not a link (it doesn't take you to a separate page, it's a button that does something within the same page). The ticket needs to be updated.
